### ISSUE TITLE
Retinanet is cpu bound

### ISSUE THREAD

**Issue by hazemawadalla** [mlcommons/storage#490]
On my testing with retinanet on a gen5 server (24c/48t), 256GB DDR5

 ┌─────┬──────┬────────────────┬────────────────┬────────────────────┐
  │ -na │ AU%  │ peak disk MB/s │ min CPU idle % │ disk % of ~13 GB/s │
  ├─────┼──────┼────────────────┼────────────────┼────────────────────┤
  │ 2   │ 97.7 │ 308            │ 90             │ 2%                 │
  ├─────┼──────┼────────────────┼────────────────┼────────────────────┤
  │ 8   │ 96.4 │ 1,220          │ 63             │ 9%                 │
  ├─────┼──────┼────────────────┼────────────────┼────────────────────┤
  │ 16  │ 93.9 │ 2,546          │ 50             │ 20%                │
  ├─────┼──────┼────────────────┼────────────────┼────────────────────┤
  │ 24  │ 90.0 │ 3,754          │ 24             │ 29%                │
  ├─────┼──────┼────────────────┼────────────────┼────────────────────┤
  │ 32  │ 85.5 │ 4,647          │ 0.00           │ 36%                │
  ├─────┼──────┼────────────────┼────────────────┼────────────────────┤
  │ 48  │ ?    │ 112            │ 0              │ — (failed)         │
  ├─────┼──────┼────────────────┼────────────────┼────────────────────┤
  │ 64  │ ?    │ 112            │ 0              │ — (failed)         │
  └─────┴──────┴────────────────┴────────────────┴────────────────────┘

retinanet's max-passing point on this host (32 accelerators, AU 85.5%) is reached when CPU idle = 0% while the Gen5 drive is only 33% utilized (4.65 of ~14 GB/s).   The cliff is the CPU, not the SSD; the same drive in a higher-core host would score a higher 'max accelerators.' retinanet's training result is therefore a CPU/host metric, not a storage one. The limiter is host CPU, not storage  a higher-core host would push the 'edge' higher with the same drive. The training-AU result characterizes the client, not the SSD.

**Comment by hazemawadalla** [comment]
removed the p0, I think the training run is useful if multihost scaling  but cpu bound on a single host.

**Comment by dslik** [comment]
Can you paste in the training datagen and run commands and I can try to reproduce this on my machine here.

**Comment by hazemawadalla** [comment]
mlpstorage closed training retinanet datasize -at b200 -cm 251 -ma 96 --hosts localhost -nc 1 --skip-validation

mlpstorage closed training retinanet datagen file --hosts 127.0.0.1 --exec-type=mpi --num-processes=8 --results-dir=/home/sped/retinanet-test --data-dir=/mnt/retinanet-test --params dataset.num_files_train=4172524 dataset.num_subfolders_train=0

for N in 1 2 8 16 24 48 64; do ( mpstat 2 > cpu_$N.txt & iostat -xm 2 /dev/nvme9n1 > io_$N.txt & mlpstorage closed training retinanet run file --hosts localhost -nc 1 -cm 251 -na $N -at b200 --results-dir /home/sped/retinanet-test --data-dir /mnt/retinanet-test --params reader.read_threads=2 ; kill %1 %2 );done

**Comment by hazemawadalla** [comment]
#!/usr/bin/env bash
  set -u
  DATADIR=/mnt/retinanet-test
  RESULTS=/home/sped/retinanet-test
  SRC=$(findmnt -no SOURCE "$DATADIR")              # real backing device, e.g. /dev/nvme9n1
  echo "data=$DATADIR  device=$SRC"
  SUM=ret_sweep_summary.tsv
  printf "accel\trthreads\tAU%%\tIO_MiBps\tpeak_rMBps\tmin_cpu_idle%%\n" > "$SUM"

  for N in 1 2 4 8 16 24 32 48 64 96; do
    sync; sudo sh -c 'echo 3 > /proc/sys/vm/drop_caches'      # cold start (also frees RAM for the worker budget)
    AVAIL=$(free -g | awk '/^Mem:/{print $7}')
    RT=$(awk -v a="$AVAIL" -v n="$N" 'BEGIN{r=int(a*0.75/(n*0.5)); if(r>8)r=8; if(r<2)r=2; print r}')  # ~0.5GB/worker, fit RAM, cap 8
    echo "=== -na $N  read_threads=$RT  (avail ${AVAIL}GB) ==="
    mpstat 2 > "cpu_$N.txt" 2>&1 &  MP=$!
    iostat -xm 2 "$SRC" > "io_$N.txt" 2>&1 &  IO=$!
    mlpstorage closed training retinanet run file --hosts localhost -nc 1 -cm 251 \
        -na "$N" -at b200 --results-dir "$RESULTS" --data-dir "$DATADIR" \
        --params train.epochs=1 reader.read_threads="$RT" > "run_$N.log" 2>&1
    kill "$MP" "$IO" 2>/dev/null

    AU=$(grep -aoE 'Accelerator Utilization \[AU\] \(%\): [0-9.]+' "run_$N.log" | tail -1 | grep -oE '[0-9.]+$')
    IOT=$(grep -aoE 'I/O Throughput \(MiB/second\): [0-9.]+' "run_$N.log" | tail -1 | grep -oE '[0-9.]+$')
    PK=$(awk -v d="${SRC##*/}" '$0 ~ d {print $3}' "io_$N.txt" | sort -n | tail -1)
    IDLE=$(awk '/all/{print $NF}' "cpu_$N.txt" | sort -n | head -1)
    printf "%s\t%s\t%s\t%s\t%s\t%s\n" "$N" "$RT" "${AU:-?}" "${IOT:-?}" "${PK:-?}" "${IDLE:-?}" | tee -a "$SUM"
  done
  echo "=== summary ==="; column -t "$SUM"

~

**Comment by russfellows** [comment]
I would say yes, this is accurate.  Retianet IS CPU bound, but it is still a decent test.  So, on a single system this is not a great storage test, but should work fine for testing a storage system, just need to use enough clients.
